_source/_posts/2019-12-12-tutorial-kotlin-beginners-guide.md

```diff
@@ -0,0 +1,491 @@
+
+```kotlin
+package com.okta.developer
+
+fun main() {
+
+    val a = arrayOf("uno", "dos", "tres")
+    readMethod(a)
+}
+
+fun readMethod(a: Array<out Any>) {
+    println(a[0])
+    a.set(0, 1) // Out-projected type prohibits the use of fun set
+}
+```
+
+The example above does not compile in Kotlin, because the `readMethod` receives the out-projected array (only read operations allowed) and it is calling the write operation `set`.
+
+**No checked exceptions**: Java checked exceptions must be somehow handled for the program to compile, and are many times swallowed by an empty catch block. Kotlin does not have checked exceptions, because it is suggested that in large software projects it decreases productivity.
+
+**Proper function types**: As opposed to Java SAM (Single Abstract Method) conversions, where a lambda expression is convertible to a SAM type according to some rules, Kotlin uses a family of function types like `(Int) -> String`, with the special notation corresponding to the signature of the functions (parameters and return values).
+
+A complete list of Kotlin features, not supported in Java, is available at [Kotlin Reference](https://kotlinlang.org/docs/reference/comparison-to-java.html#what-kotlin-has-that-java-does-not).
+
+## Java and Kotlin Interoperability
+
+Kotlin is 100% interoperable with Java, Kotlin code can be called from Java and vice versa.
+Existing Java code can be used with some considerations, for example, Java getters and setters are represented by properties in Kotlin:
+
+```kotlin
+val calendar = Calendar.getInstance()
+calendar.firstDayOfWeek = Calendar.MONDAY // setFirstOfWeek()
+```
+
+With an IDE like IntelliJ IDEA, you can add Java source code to a Kotlin project just by creating the `.java` file.
+
+Objects coming from Java, called **platform types**, have relaxed null-checks for practical reasons, and safety is the same as in Java. Kotlin will not inform a compilation error but the call might fail at runtime. For example, if Kotlin calls Java code that returns an ArrayList, the inferred type in Kotlin will be ArrayList<String!>!, which means the collection can be nullable, and the items as well.
+
+In the Java code below, `null` is added as an item to the list.
+
+```java
+import java.util.ArrayList;
+
+public class ArrayWithNulls {
+
+    public static ArrayList<String> create(){
+        ArrayList<String> strings = new ArrayList();
+        strings.add(null);
+        strings.add("foo");
+        return strings;
+    }
+}
+```
+In the Kotlin lines below, we are calling the previous Java code, so the safety is relaxed. The program compiles, but as the first item is null, it will fail at runtime with IllegalStateException. 
```

Comment:
```suggestion
In the Kotlin lines below, we are calling the previous Java code, so the safety is relaxed. The program compiles, but as the first item is `null`, it will fail at runtime with `IllegalStateException`. 
```